File: pkg/end_tests/two_player_game_test.go
package end_tests

import (
	"testing"

	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/deck"
	gameMod "github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game/elements"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/stack"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tiles/feature"
	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tiles/side"
)

/*
 diagonal edges represent cities, dots fields, straight lines roads. The big vertical line on the left is to prevent comment formating

 Final board: (each tile is repestned by 3x3 ascii signs, at the center is the turn number in hex :/)

| 	.|........|.
| 	.9--1--2..8.
| 	.|./ \ |..|.
| 	.| \ / |..|....
| 	.4--0--3..B--C-
| 	.|.............
|	.|.   .........
| 	.5.   >6<>7<>A<
| 	...   .........
*/

func TestFullGame(t *testing.T) {
	// create game
	minitileSet := MiniTileSetRoadsAndFields()
	deckStack := stack.NewSeeded(minitileSet.Tiles, 47328235) // random seed :P
	deck := deck.Deck{Stack: &deckStack, StartingTile: minitileSet.StartingTile}
	game, err := gameMod.NewFromDeck(deck, nil)
	if err != nil {
		t.Fatal(err.Error())
	}
	/*
		for i, _ := range deck.GetTiles() {
			fmt.Printf("%v:\n", i+1)
			var tile, _ = deck.Next()
			fmt.Printf("%v\n", tile)
		}
	*/

	checkFirstTurn(game, t)    // straight road with city edge
	checkSecondTurn(game, t)   // road turn
	checkThirdTurn(game, t)    // road turn
	checkFourthTurn(game, t)   // T cross road
	checkFifthTurn(game, t)    // monastery with single road
	checkSixthTurn(game, t)    // Two city edges not connected
	checkSeventhTurn(game, t)  // Two city edges not connected
	checkEightthTurn(game, t)  // straight road
	checkNinethTurn(game, t)   // T cross road
	checkTenthTurn(game, t)    // Two city edges not connected
	checkEleventhTurn(game, t) // road turn
	checkTwelvethTurn(game, t) // straight road
	checkFinalResult(game, t)

}

func MakeTurn(game *gameMod.Game, t *testing.T, tilePosition elements.Position, rotations uint, meeple elements.MeepleType, featureSide side.Side, featureType feature.Type) {
	tile, err := game.GetCurrentTile()
	if err != nil {
		t.Fatal(err.Error())
	}

	var player = game.CurrentPlayer()

	ptile := elements.ToPlacedTile(tile.Rotate(rotations))
	ptile.Position = tilePosition
	if meeple != elements.NoneMeeple {
		ptile.GetPlacedFeatureAtSide(featureSide, featureType).Meeple = elements.Meeple{
			MeepleType: meeple,
			PlayerID:   player.ID(),
		}
	}

	err = game.PlayTurn(ptile)
	if err != nil {
		t.Fatal(err.Error())
	}
}

func CheckMeeplesAndScore(game *gameMod.Game, t *testing.T, player1Score uint32, player1Meeples uint8, player2Score uint32, player2Meeples uint8) {
	var player1 = game.GetPlayerByID(1)
	var player2 = game.GetPlayerByID(2)

	// player 1

	// check meeples count of first player
	if player1.MeepleCount(elements.NormalMeeple) == player1Meeples {
		t.Fatalf("meeples count does not match. Expected: %d  Got: %d", player1Meeples, player1.MeepleCount(elements.NormalMeeple))
	}
	// check points
	if player1.Score() != player1Score {
		t.Fatalf("Player received wrong amount of points! Expected: %d  Got: %d ", player1Score, player1.Score())
	}

	// player 2

	// check meeples count of first player
	if player2.MeepleCount(elements.NormalMeeple) == player2Meeples {
		t.Fatalf("meeples count does not match. Expected: %d  Got: %d", player2Meeples, player2.MeepleCount(elements.NormalMeeple))
	}
	// check points
	if player2.Score() != player2Score {
		t.Fatalf("Player received wrong amount of points! Expected: %d  Got: %d ", player2Score, player2.Score())
	}
}

// straight road with city edge
// player 1 places meeple on city, and closes it
/*
|       0
|
| 	   ...
|1 	   -1-
| 	   / \
| 	   \ /
|0 	   -0-
| 	   ...
*/
func checkFirstTurn(game *gameMod.Game, t *testing.T) {

	MakeTurn(game, t, elements.NewPosition(0, 1), 2, elements.NormalMeeple, side.Bottom, feature.City)
	CheckMeeplesAndScore(game, t, 4, 7, 0, 7)
}

// road turn
// player 2 places meeple (M) on a road
/*
|       0  1
|
| 	   ......
|1 	   -1-M2.
| 	   / \ |.
| 	   \ /
|0 	   -0-
| 	   ...
*/
func checkSecondTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(1, 1), 0, elements.NormalMeeple, side.Left, feature.Road)
	CheckMeeplesAndScore(game, t, 4, 7, 0, 6)
}

// road turn
// player 1 places meeple (m) on a field
/*
|       0  1
|
| 	   ......
|1 	   -1-M2.
| 	   / \ |.
| 	   \ /m|.
|0 	   -0--3.
| 	   ......
*/
func checkThirdTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(1, 0), 0, elements.NoneMeeple, side.None, feature.None)
	CheckMeeplesAndScore(game, t, 4, 6, 0, 6)
}

// T cross road
// player2 places meeple on road going down
/*
|   -1  0  1
|
| 	   ......
|1 	   -1-M2.
| 	   / \ |.
| 	.| \ /m|.
|0 	.4--0--3.
| 	.M.......
*/
func checkFourthTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(-1, 0), 0, elements.NormalMeeple, side.Bottom, feature.Road)
	CheckMeeplesAndScore(game, t, 4, 6, 0, 5)
}

// monastery with single road
// player1 places meeple on a monastery
// road from 4 to 5 is finished, so player2 scores 2 points
/*
|   -1  0  1
|
| 	   ......
|1 	   -1-M2.
| 	   / \ |.
| 	.| \ /m|.
|0 	.4--0--3.
| 	.|.......
|	.|.
|-1 .5.
| 	...
*/
func checkFifthTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(-1, 0), 0, elements.NormalMeeple, side.None, feature.Monastery)
	CheckMeeplesAndScore(game, t, 4, 5, 2, 6)

}

// Two city edges not connected
// player 2 places meeple(M) on the right city
/*
|   -1  0  1
|
| 	   ......
|1 	   -1-M2.
| 	   / \ |.
| 	.| \ /m|.
|0 	.4--0--3.
| 	.|.......
|	.|.   ...
|-1 .5.   >6<M
| 	...   ...
*/
func checkSixthTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(1, -1), 0, elements.NormalMeeple, side.Right, feature.City)
	CheckMeeplesAndScore(game, t, 4, 5, 2, 5)
}

// Two city edges not connected
// player 1 places meeple on the right city
// playey 2 scores points for finished city

/*
|   -1  0  1  2
|
| 	   ......
|1 	   -1-M2.
| 	   / \ |.
| 	.| \ /m|.
|0 	.4--0--3.
| 	.|.......
|	.|.   ......
|-1	.5.   >6<>7<m
| 	...   ......
*/
func checkSeventhTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(2, -1), 0, elements.NormalMeeple, side.Right, feature.City)
	CheckMeeplesAndScore(game, t, 4, 4, 6, 6)
}

// straight road
// player 2 places meeple on a bottom road

/*
|   -1  0  1  2  3
|
| 	   .......|.
|1 	   -1-M2..8.
| 	   / \ |..M.
| 	.| \ /m|.
|0 	.4--0--3.
| 	.|.......
|	.|.   ......
|-1 .5.   >6<>7<m
| 	...   ......
*/
func checkEightthTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(1, 1), 0, elements.NormalMeeple, side.Bottom, feature.Road)
	CheckMeeplesAndScore(game, t, 4, 4, 6, 5)
}

// T cross road
// road is finished. Player 2 scores 6 points for a road
// player 1 places meeple (m) on a field

/*
|   -1  0  1  2  3
|
| 	.|m.......|.
|1 	.9--1--2..8.
| 	.|./ \ |..M.
| 	.| \ /m|.
|0 	.4--0--3.
| 	.|.......
|	.|.   ......
|-1	.5.   >6<>7<m
| 	...   ......
*/
func checkNinethTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(-1, 1), 0, elements.NormalMeeple, side.TopRightEdge, feature.Field)
	CheckMeeplesAndScore(game, t, 4, 3, 12, 6)
}

// Two city edges not connected
// player 2 places meeple (M) on the right city
// player 1 scores points for city

/*
|   -1  0  1  2  3
|
| 	.|m.......|.
|1 	.9--1--2..8.
| 	.|./ \ |..M.
| 	.| \ /m|.
|0 	.4--0--3.
| 	.|.......
|	.|.   .........
|-1	.5.   >6<>7<>A<M
| 	...   .........
*/
func checkTenthTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(3, -1), 0, elements.NormalMeeple, side.Right, feature.City)
	CheckMeeplesAndScore(game, t, 8, 4, 12, 5)
}

// road turn

/*
|   -1  0  1  2  3
|
| 	.|m.......|.
|1 	.9--1--2..8.
| 	.|./ \ |..M.
| 	.| \ /m|..|.
|0 	.4--0--3..B-
| 	.|..........
|	.|.   .........
|-1	.5.   >6<>7<>A<M
| 	...   .........
*/
func checkEleventhTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(2, 0), 0, elements.NoneMeeple, side.None, feature.None)
	CheckMeeplesAndScore(game, t, 8, 4, 12, 5)
}

// straight road
/*
|   -1  0  1  2  3
|
| 	.|m.......|.
|1 	.9--1--2..8.
| 	.|./ \ |..M.
| 	.| \ /m|..|....
|0 	.4--0--3..B--C-
| 	.|.............
|	.|.   .........
|-1	.5.   >6<>7<>A<M
| 	...   .........
*/
func checkTwelvethTurn(game *gameMod.Game, t *testing.T) {
	MakeTurn(game, t, elements.NewPosition(3, 0), 0, elements.NoneMeeple, side.None, feature.None)
	CheckMeeplesAndScore(game, t, 8, 4, 12, 5)
}

/* player 1 scores additional 17 points:
	- 2 points for monastery
	- 6 points for farmer in the center
	- 9 points for farmer outside
   player 2 scores additional 4 points:
	- 3 points for a road
	- 1 point for unfishied city
*/

func checkFinalResult(game *gameMod.Game, t *testing.T) {
	var scores, err = game.Finalize()
	if err != nil {
		t.Fatal(err.Error())
	}
	if scores[0] != 25 {
		t.Fatal("Player 1 final score incorrect. Expected 25, got: %d", scores[1])
	}

	if scores[1] != 16 {
		t.Fatal("Player 2 final score incorrect. Expected 16, got: %d", scores[1])
	}
}

File: python_bindings/tests/end-to-python-end-test/two_player_game_test.py
# flake8: noqa ascci drawing makes a lot errors
# mypy: ignore-errors

import logging
from pathlib import Path

from end_utils import TurnParams, check_points, make_turn

from carcassonne_engine import GameEngine, tiletemplates
from carcassonne_engine._bindings.elements import MeepleType
from carcassonne_engine._bindings.feature import Type as FeatureType
from carcassonne_engine._bindings.side import Side
from carcassonne_engine.models import SerializedGame
from carcassonne_engine.placed_tile import Position
from carcassonne_engine.tilesets import TileSet

"""
 diagonal edges represent cities, dots fields, straight lines roads.
 Final board: (each tile is represented by 5x5 ascii signs, at the center is the turn number in hex :/)

 
            -1    0    1    2    3
            
            
           ..|..............|..     
           ..|..............|..     
1          ..9----1----2....8..     
           ..|.../ \...|....|..     
           ..|..|   |..|....|..     
           ..|..|   |..|....|.......
           ..|...\ /...|....|.......
0          ..4----0----3....B----C--
           ..|......................
           ..|......................
           ..|..     -...--...--...-
           .[ ].      \./  \./  \./
-1         .[5].       6    7    A
           .[ ].      /.\  /.\  /.\ 
           .....     -...--   --...-
"""

log = logging.getLogger(__name__)


def test_two_player_game(tmp_path: Path) -> None:
    engine = GameEngine(4, tmp_path)
    tile_set = create_tileset()

    game_id, game = engine.generate_ordered_game(tile_set)

    game_id, game = check_first_turn(engine, game, game_id)
    game_id, game = check_second_turn(engine, game, game_id)
    game_id, game = check_third_turn(engine, game, game_id)
    game_id, game = check_fourth_turn(engine, game, game_id)
    game_id, game = check_fifth_turn(engine, game, game_id)
    game_id, game = check_sixth_turn(engine, game, game_id)
    game_id, game = check_seventh_turn(engine, game, game_id)
    game_id, game = check_eighth_turn(engine, game, game_id)
    game_id, game = check_ninth_turn(engine, game, game_id)
    game_id, game = check_tenth_turn(engine, game, game_id)
    game_id, game = check_eleventh_turn(engine, game, game_id)
    game_id, game = check_twelfth_turn(engine, game, game_id)

    assert game.current_tile is None


def create_tileset() -> TileSet:
    tiles = [
        tiletemplates.single_city_edge_straight_roads().rotate(2),
        tiletemplates.roads_turn(),
        tiletemplates.roads_turn().rotate(1),
        tiletemplates.t_cross_road().rotate(3),
        tiletemplates.monastery_with_single_road().rotate(2),
        tiletemplates.two_city_edges_up_and_down_not_connected().rotate(1),
        tiletemplates.two_city_edges_up_and_down_not_connected().rotate(1),
        tiletemplates.straight_roads().rotate(1),
        tiletemplates.t_cross_road().rotate(3),
        tiletemplates.two_city_edges_up_and_down_not_connected().rotate(1),
        tiletemplates.roads_turn().rotate(2),
        tiletemplates.straight_roads(),
    ]

    return TileSet.from_tiles(
        tiles, starting_tile=tiletemplates.single_city_edge_straight_roads()
    )


"""
// straight road with city edge
// player1 places meeple on city, and closes it
// player 1 scores 4 points

            -1    0    1    2    3
            
            
                .....               
                .....               
1               --1--              
                ./ \.               
                |   |               
                |   |               
                .\ /.               
0               --0--               
                .....               
                .....               
                                    
                                   
-1                                
                                    
 
"""


def check_first_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=0, y=1),
        tile=game.current_tile,
        meepleType=MeepleType.NormalMeeple,
        side=Side.Bottom,
        featureType=FeatureType.City,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [0 + 4, 0])

    return game_id, game


"""
// road turn
// player2 places meeple (@) on a road
            -1    0    1    2    3
            
            
                ..........          
                ..........          
1               --1--@-2..         
                ./ \...|..          
                |   |..|..          
                |   |               
                .\ /.               
0               --0--               
                .....               
                .....               
                                    
                                   
-1                                
                                    
  
"""


def check_second_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=1, y=1),
        tile=tiletemplates.roads_turn(),
        meepleType=MeepleType.NormalMeeple,
        side=Side.Left,
        featureType=FeatureType.Road,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [4, 0])

    return game_id, game


"""
// road turn
// player1 places meeple (!) on a field
            -1    0    1    2    3
            
            
                ..........          
                ..........          
1               --1--@-2..         
                ./ \...|..          
                |   |..|..          
                |   |..|..          
                .\ /.!.|..          
0               --0----3..          
                ..........          
                ..........          
                                    
                                   
-1                                
                                    
    
"""


def check_third_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=1, y=0),
        tile=game.current_tile,
        meepleType=MeepleType.NormalMeeple,
        side=Side.TopLeftEdge,
        featureType=FeatureType.Field,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [4, 0])

    return game_id, game


"""
// T cross road
// player2 places meeple (@) on road going down
            -1    0    1    2    3
            
            
                ..........          
                ..........          
1               --1--@-2..         
                ./ \...|..          
                |   |..|..          
           ..|..|   |..|..          
           ..|...\ /.!.|..          
0          ..4----0----3..          
           ..|............          
           ..@............          
                                    
                                   
-1                                
                                    
                        
"""


def check_fourth_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=-1, y=0),
        tile=game.current_tile,
        meepleType=MeepleType.NormalMeeple,
        side=Side.Bottom,
        featureType=FeatureType.Road,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [4, 0])

    return game_id, game


"""
// monastery with single road
// player1 places meeple (!) on a monastery
// road from 4 to 5 is finished, so player2 scores 2 points
            -1    0    1    2    3
            
            
                ..........          
                ..........          
1               --1--@-2..         
                ./ \...|..          
                |   |..|..          
           ..|..|   |..|..          
           ..|...\ /.!.|..          
0          ..4----0----3..          
           ..|............          
           ..|............          
           ..|..                    
           .[ ].                   
-1         .[5].                  
           .[!].                    
           .....            
"""


def check_fifth_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=-1, y=-1),
        tile=game.current_tile,
        meepleType=MeepleType.NormalMeeple,
        side=Side.NoSide,
        featureType=FeatureType.Monastery,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [4, 0 + 2])

    return game_id, game


"""
// Two city edges not connected
// player2 places meeple (@) on the right city
            -1    0    1    2    3
            
            
                ..........          
                ..........          
1               --1--@-2..         
                ./ \...|..          
                |   |..|..          
           ..|..|   |..|..          
           ..|...\ /.!.|..          
0          ..4----0----3..          
           ..|............          
           ..|............          
           ..|..     -...-          
           .[ ].      \./          
-1         .[5].       6 @        
           .[!].      /.\           
           .....     -...-      
"""


def check_sixth_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=1, y=-1),
        tile=game.current_tile,
        meepleType=MeepleType.NormalMeeple,
        side=Side.Right,
        featureType=FeatureType.City,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [4, 2])

    return game_id, game


"""
// Two city edges not connected
// player1 places meeple (!) on the right city
// player2 scores points for finished city
            -1    0    1    2    3
            
            
                ..........          
                ..........          
1               --1--@-2..         
                ./ \...|..          
                |   |..|..          
           ..|..|   |..|..          
           ..|...\ /.!.|..          
0          ..4----0----3..          
           ..|............          
           ..|............          
           ..|..     -...--...-     
           .[ ].      \./  \./     
-1         .[5].       6    7 !   
           .[!].      /.\  /.\      
           .....     -...--...- 
"""


def check_seventh_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=2, y=-1),
        tile=game.current_tile,
        meepleType=MeepleType.NormalMeeple,
        side=Side.Right,
        featureType=FeatureType.City,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [4, 2 + 4])

    return game_id, game


"""
// straight road
// player2 places meeple (@) on the road
            -1    0    1    2    3
            
            
                ............|..     
                ............|..     
1               --1--@-2....8..     
                ./ \...|....|..     
                |   |..|....@..     
           ..|..|   |..|..          
           ..|...\ /.!.|..          
0          ..4----0----3..          
           ..|............          
           ..|............          
           ..|..     -...--...-     
           .[ ].      \./  \./     
-1         .[5].       6    7 !   
           .[!].      /.\  /.\      
           .....     -...--...-   
"""


def check_eighth_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=2, y=1),
        tile=game.current_tile,
        meepleType=MeepleType.NormalMeeple,
        side=Side.Bottom,
        featureType=FeatureType.Road,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [4, 6])

    return game_id, game


"""
// T cross road
// road is finished. Player 2 scores 6 points for a road
// player1 places meeple (!) on a field
            -1    0    1    2    3
            
            
           ..|.!............|..     
           ..|..............|..     
1          ..9----1----2....8..     
           ..|.../ \...|....|..     
           ..|..|   |..|....@..     
           ..|..|   |..|..          
           ..|...\ /.!.|..          
0          ..4----0----3..          
           ..|............          
           ..|............          
           ..|..     -...--...-     
           .[ ].      \./  \./     
-1         .[5].       6    7 !    
           .[!].      /.\  /.\      
           .....     -...--...-     
"""


def check_ninth_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=-1, y=1),
        tile=game.current_tile,
        meepleType=MeepleType.NormalMeeple,
        side=Side.TopRightEdge,
        featureType=FeatureType.Field,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [4, 6 + 6])

    return game_id, game


"""
// Two city edges not connected
// player2 places meeple (@) on the right city
// player1 scores points for city
            -1    0    1    2    3
            
            
           ..|.!............|..     
           ..|..............|..     
1          ..9----1----2....8..     
           ..|.../ \...|....|..     
           ..|..|   |..|....@..     
           ..|..|   |..|..          
           ..|...\ /.!.|..          
0          ..4----0----3..          
           ..|............          
           ..|............          
           ..|..     -...--...--...-
           .[ ].      \./  \./  \./
-1         .[5].       6    7    A @
           .[!].      /.\  /.\  /.\ 
           .....     -...--...--...-
"""


def check_tenth_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=3, y=-1),
        tile=game.current_tile,
        meepleType=MeepleType.NormalMeeple,
        side=Side.Right,
        featureType=FeatureType.City,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [4 + 4, 12])
    return game_id, game


"""
// road turn
            -1    0    1    2    3
            
            
           ..|.!............|..     
           ..|..............|..     
1          ..9----1----2....8..     
           ..|.../ \...|....|..     
           ..|..|   |..|....@..     
           ..|..|   |..|....|..     
           ..|...\ /.!.|....|..     
0          ..4----0----3....B--     
           ..|.................     
           ..|.................     
           ..|..     -...--...--...-
           .[ ].      \./  \./  \./
-1         .[5].       6    7    A @
           .[!].      /.\  /.\  /.\ 
           .....     -...--...--...-
"""


def check_eleventh_turn(
    engine: GameEngine, game, game_id
) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=2, y=0),
        tile=game.current_tile,
        meepleType=MeepleType.NoneMeeple,
        side=Side.NoSide,
        featureType=FeatureType.NoneType,
    )

    game_id, game = make_turn(engine, game, game_id, turn_params)
    check_points(game, [8, 12])

    return game_id, game


"""
// straight road
            -1    0    1    2    3
            
            
           ..|.!............|..     
           ..|..............|..     
1          ..9----1----2....8..     
           ..|.../ \...|....|..     
           ..|..|   |..|....@..     
           ..|..|   |..|....|.......
           ..|...\ /.!.|....|.......
0          ..4----0----3....B----C--
           ..|......................
           ..|......................
           ..|..     -...--...--...-
           .[ ].      \|/  \./  \./
-1         .[5].       6    7    A @
           .[!].      /.\  /.\  /.\ 
           .....     -...--...--...-
"""


def check_twelfth_turn(engine: GameEngine, game, game_id) -> tuple[int, SerializedGame]:
    turn_params = TurnParams(
        pos=Position(x=3, y=0),
        tile=tiletemplates.straight_roads(),
        meepleType=MeepleType.NoneMeeple,
        side=Side.NoSide,
        featureType=FeatureType.NoneType,
    )

    game_id, game = make_turn(
        engine, game, game_id, turn_params, {1: 8 + 12, 2: 12 + 4}
    )
    check_points(game, [8, 12])

    return game_id, game


"""
player1 scores 11 points in total:
    - 2*3points for farmer on 9 tile
    - 3 points for farmer on 3 tile
    - 3 points for monastery on 5 tile  

player2 scores 4 points in total:
    - 3 points for road on 8 tile
    - 1 point for city on A tile    

            -1    0    1    2    3


           ..|.!............|..     
           ..|..............|..     
1          ..9----1----2....8..     
           ..|.../ \...|....|..     
           ..|..|   |..|....@..     
           ..|..|   |..|....|.......
           ..|...\ /.!.|....|.......
0          ..4----0----3....B----C--
           ..|......................
           ..|......................
           ..|..     -...--...--...-
           .[ ].      \|/  \./  \./
-1         .[5].       6    7    A @
           .[!].      /.\  /.\  /.\ 
           .....     -...--...--...-
"""
